### PR TITLE
ARROW-12378: [C++][Gandiva] Implement castVARBINARY functions

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement the following functions:

- castVARBINARY(varchar, len)
- castVARBINARY(varbinary, len)
- castVARBINARY(float/double/int/bigint, bigint len)

### CODE REVIEW

File: cpp/src/gandiva/context_helper.cc [hunk 1]
```diff
@@ -69,6 +69,11 @@ uint8_t* gdv_fn_context_arena_malloc(int64_t context_ptr, int32_t size) {
   return context->arena()->Allocate(size);
 }
 
+uint8_t* gdv_fn_context_arena_malloc_int64(int64_t context_ptr, int64_t size) {
```
Comment (projjal) [inline cpp/src/gandiva/context_helper.cc:72]: why not keep only one method.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 2]
```diff
@@ -216,6 +216,72 @@ SHA_VAR_LEN_PARAMS(SHA1_HASH_FUNCTION_BUF)
...
+
+CAST_VARBINARY_FROM_STRING_AND_BINARY(utf8)
+CAST_VARBINARY_FROM_STRING_AND_BINARY(binary)
+
+#undef CAST_VARBINARY_FROM_STRING_AND_BINARY
+
+// Add functions for castVARBINARY for numeric
+#define CAST_VARBINARY_FROM_NUMERIC(TYPE_NAME)                                         \
+  GANDIVA_EXPORT                                                                       \
+  const char* gdv_fn_castVARBINARY_##TYPE_NAME(int64_t context, gdv_##TYPE_NAME value, \
```
Comment (projjal) [inline cpp/src/gandiva/gdv_function_stubs.cc:247]: I think you should return the the bytes corresponding to the string representation of int/float. (opposite of what the change castINT from varbinary is doing)

Comment (jpedroantunes) [reply cpp/src/gandiva/gdv_function_stubs.cc:250]: If I get it correctly, I should do the same as the castVARCHAR for numeric types, but returning the correspondent bytes, correct?

Comment (projjal) [reply cpp/src/gandiva/gdv_function_stubs.cc:250]: yes..you can resuse that macro i suppose

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 3]
```diff
@@ -307,85 +307,114 @@ CAST_NUMERIC_FROM_STRING(double, arrow::DoubleType, FLOAT8)
...
-    });                                                                                  \
-    if (!status.ok()) {                                                                  \
-      std::string err = "Could not cast " + std::to_string(value) + " to string";        \
-      gdv_fn_context_set_error_msg(context, err.c_str());                                \
-      *out_len = 0;                                                                      \
-      return "";                                                                         \
-    }                                                                                    \
-    return ret;                                                                          \
+// Add functions for castVARBINARY for utf8 and binary
+#define CAST_VARBINARY_FROM_STRING_AND_BINARY(TYPE_NAME)                               \
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Solved

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 4]
```diff
@@ -307,85 +307,114 @@ CAST_NUMERIC_FROM_STRING(double, arrow::DoubleType, FLOAT8)
...
+      gdv_fn_context_set_error_msg(context, err.c_str());                         \
+      *out_len = 0;                                                               \
+      return "";                                                                  \
+    }                                                                             \
+    return ret;                                                                   \
+  }
 
-#undef GDV_FN_CAST_VARCHAR_INTEGER
-#undef GDV_FN_CAST_VARCHAR_REAL
+GDV_FN_CAST_VARCHAR_VARBINARY_INTEGER(int32, VARCHAR, Int32Type)
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Solved

File: cpp/src/gandiva/function_registry_string.cc [hunk 5]
```diff
@@ -236,6 +236,30 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
       NativeFunction("binary_string", {}, DataTypeVector{utf8()}, binary(),
                      kResultNullIfNull, "binary_string", NativeFunction::kNeedsContext),
 
+      NativeFunction("castVARBINARY", {}, DataTypeVector{binary(), int64()}, binary(),
+                     kResultNullIfNull, "gdv_fn_castVARBINARY_binary",
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/function_registry_string.cc]: Solved

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 6]
```diff
@@ -533,89 +533,102 @@ const char* castVARCHAR_bool_int64(gdv_int64 context, gdv_boolean value,
...
+    if (byte_pos < 0) {                                                                \
+      *out_length = 0;                                                                 \
+      return "";                                                                       \
+    }                                                                                  \
+                                                                                       \
+    *out_length = index + byte_pos;                                                    \
+    return data;                                                                       \
+  }
+
+CAST_FROM_STRING_AND_BINARY(VARCHAR, utf8)
```
[the remark this answers is not in the document]
Comment (jpedroantunes) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Ok! Used the implementation I did before

**Review by projjal — APPROVED** [review]

**Comment by praveenbingo** [comment]
@projjal can you please check the errors on ths one.

**Comment by projjal** [comment]
@jpedroantunes this needs a rebase

**Pull request closed without merging**